File: days/day_01/main.ml
open Core

let file = "input.txt"

let part_one nums =
	let num =
	 	List.find nums ~f:(fun x -> 
			List.exists nums ~f:(fun y -> y = 2020 - x)
		)
	in
	match num with
	| Some y -> y*(2020-y)
	| None -> 0
;;

let part_two nums =
	let num =
	 	List.find_map nums ~f:(fun x -> 
			let y = List.find_map nums ~f:(fun z -> 
				List.find nums ~f:(fun w -> w = 2020 - x - z)
			)
			in
			match y with
		 	| Some y -> Some (x*y*(2020-x-y))
		 	| None -> None
	 	)
	in
	match num with
	| Some num -> num
	| None -> 0
;;

let () = 
	let start_time = Unix.gettimeofday () in
	let input = In_channel.read_lines file in
	let nums = List.map input ~f:int_of_string in
	let part_one_ans = part_one nums in
	let part_two_ans = part_two nums in
	printf "Day  1 (%.3fs)\n" ((Unix.gettimeofday ()) -. start_time);
	printf "  Part 1: %d\n" part_one_ans;
	printf "  Part 2: %d\n" part_two_ans
;;

File: days/day_01/main.py
import time
import os

def part_one(nums):
	for x in nums:
		for y in nums:
			if x + y == 2020:
				return x * y

				
def part_two(nums):
	for x in nums:
		for y in nums:
			for z in nums:
				if x + y + z == 2020:
					return x * y * z

					
def main():
	start_time = time.time()

	with open(os.path.dirname(__file__) + '/input.txt', 'r') as data:
		nums = [int(line) for line in data.readlines()]
		
		part_one_ans = part_one(nums)
		part_two_ans = part_two(nums)

		print('Day  1 ({:,.3f}s)'.format(time.time()-start_time))
		print('  Part 1: {}'.format(part_one_ans))
		print('  Part 2: {}'.format(part_two_ans))
		

if __name__ == "__main__":
		main()